### PR TITLE
operator-client bi-direction event system in WebSocket

### PR DESCRIPTION
**Pull request by zhouer**
When sending transactions, the receiver client needs readltime
notifications because it needs to confirm that the history has been
received. And due to the clients may not always have public ip availability,
it's easier to set a persistent connection between the server and a client,
and let the server push notifications to clients.

Now, the child_chain_client connects to the server in its constructor,
and provides `emit`, `on` event system for easier communications.
Also, the client can register its identity, like address, after connecting
to the server, and the server can identify each client to send or relay
messages. For example:

Alice:
```
c = container.get_child_chain_client()
c.emit('join', 'alice')
c.on('relay', print)
```

Bob:
```
c = container.get_child_chain_client()
c.emit('join', 'bob')
c.emit('relay', {'dest': 'alice', 'message': 'hello'})
```

And due to a WebSocket framework was introduced on server side,
the server startup command was changed to
```
python -m plasma_cash.child_chain
```
the orignal `flask run` command will failed to start the WebSocket
service.

### CODE REVIEW

File: plasma_cash/child_chain/server.py [hunk 1]
```diff
@@ -32,3 +38,22 @@ def submit_block():
...
+    global clients
+    try:
+        while True:
+            data = json.loads(socket.receive())
+
+            if (data['event'] == 'join'):
+                clients[data['arg']] = socket
+            elif(data['event'] == 'left'):
+                del clients[data['arg']]
+            elif (data['event'] == 'relay'):
```
Comment (bun919tw) [inline plasma_cash/child_chain/server.py]: Why need parentheses?

**Comment by zhouer** [comment]
Hi @bun919tw 
I fix the parentheses issue (and re-write the git history for some reasons), please review again, thanks!

File: plasma_cash/child_chain/server.py [hunk 2]
```diff
@@ -32,3 +38,22 @@ def submit_block():
 def send_tx():
     tx = request.form['tx']
     return container.get_child_chain().apply_transaction(tx)
+
+
+@ws.route('/')
+def websocket(socket):
+    global clients
```
Comment (boolafish) [inline plasma_cash/child_chain/server.py]: can we just put clients here?

File: plasma_cash/client/child_chain_client.py [hunk 3]
```diff
@@ -1,15 +1,34 @@
...
+    def __init__(self, base_url, ws_url, verify=False, timeout=5):
         self.base_url = base_url
         self.verify = verify
         self.timeout = timeout
 
+        self.ws = websocket.WebSocketApp(ws_url, on_message=self.ws_on_message)
+        self.ws_callback = {}
+        threading.Thread(target=self.ws.run_forever).start()
+
+    def ws_on_message(self, ws, message):
```
Comment (boolafish) [inline plasma_cash/client/child_chain_client.py]: if possible please add unit test for these new functions as well

File: plasma_cash/child_chain/server.py (hunk 2 above)
Comment (zhouer) [reply plasma_cash/child_chain/server.py]: Considering the client notification may not always triggered from WebSocket (but from other regular API, eg. `send_tx`), I think the `clients` need to be a global variable.

File: plasma_cash/client/child_chain_client.py (hunk 3 above)
Comment (zhouer) [reply plasma_cash/client/child_chain_client.py]: sure.

**Comment by zhouer** [comment]
Flask-SocketIO got much more attentions, adoptions, and well maintained.
SocketIO uses WebSocket as "one of its persistent connection solutions", but not uses WebSocket exclusively.
To me, WebSocket *is* a W3C standard, but SocketIO *is not* a standard, and I preferred standardized protocols better. 
However, I can always switch to SocketIO, what do you think.

**Comment by zhouer** [comment]
Flask-SocketIO overwrites flask command `flask run`
https://github.com/miguelgrinberg/Flask-SocketIO/blob/master/setup.py#L27
https://dormousehole.readthedocs.io/en/latest/cli.html#id11

**Comment by zhouer** [comment]
Hi @boolafish
1. The Flask-Sockets has been removed
2. Operator server-side event handling is using SocketIO-style decorator functions
3. child_chain_client unit tests for WebSocket related methods have been added
4. `flask run` is still **not** supported

**Comment by zhouer** [comment]
I'm not sure what functions will be needed for implementing plasma-cash, functions currently supported are more like proof of concepts. I will leave the tests in the future PR.

**Pull request merged**